File: src/main/java/com/braydenoneal/blang/parser/expression/value/Value.java
package com.braydenoneal.blang.parser.expression.value;

import com.braydenoneal.blang.parser.expression.Expression;
import com.mojang.serialization.Codec;

public abstract class Value<T> implements Expression {
    private T value;

    public Value(T value) {
        this.value = value;
    }

    public T value() {
        return value;
    }

    public void setValue(T newValue) {
        value = newValue;
    }

    @Override
    public Value<?> evaluate() {
        return this;
    }

    @Override
    public String toString() {
        return value().toString();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Value<?> value1) {
            return value.equals(value1.value());
        }

        return super.equals(obj);
    }

    public abstract ValueType<?> getType();

    private static final Codec<ValueType<?>> VALUE_TYPE_CODEC = ValueType.REGISTRY.getCodec();
    public static final Codec<Value<?>> CODEC = VALUE_TYPE_CODEC.dispatch("type", Value::getType, ValueType::codec);
}

File: src/main/java/com/braydenoneal/blang/parser/expression/value/Value.kt
package com.braydenoneal.blang.parser.expression.value

import com.braydenoneal.blang.parser.Program
import com.braydenoneal.blang.parser.expression.Expression
import com.braydenoneal.blang.parser.expression.ExpressionType
import com.braydenoneal.blang.parser.expression.ExpressionTypes
import com.mojang.serialization.Codec
import com.mojang.serialization.MapCodec

abstract class Value<T>(private val value: T) : Expression {
    fun value(): T {
        return value
    }

    override fun evaluate(program: Program): Value<*> {
        return this
    }

    override fun toString(): String {
        return value().toString()
    }

    override fun equals(other: Any?): Boolean {
        if (other is Value<*>) {
            return value == other.value()
        }

        return super.equals(other)
    }

    override val type: ExpressionType<*> get() = ExpressionTypes.VALUE

    abstract val valueType: ValueType<*>

    companion object {
        val CODEC: Codec<Value<*>> = ValueType.REGISTRY.getCodec().dispatch("type", Value<*>::valueType, ValueType<*>::codec)
        val MAP_CODEC: MapCodec<Value<*>> = CODEC.fieldOf("value")
    }

    override fun hashCode(): Int {
        var result = value?.hashCode() ?: 0
        result = 31 * result + type.hashCode()
        result = 31 * result + valueType.hashCode()
        return result
    }
}
